src/internal/ppsutil/decoder_test.go

```diff
@@ -0,0 +1,114 @@
+package ppsutil_test
+
+import (
+	"encoding/json"
+	"io"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/ppsutil"
+)
+
+func TestSpecReader(t *testing.T) {
+	var cases = map[string]struct {
+		input     string   // YAML
+		expected  []string // list of JSON specs
+		expectErr bool
+	}{
+		"empty input is okay": {},
+		"invalid JSON is bad": {
+			input:     `"dkjfskjf`,
+			expectErr: true,
+		},
+		"valid JSON is valid": {
+			input:    `{"pipeline": {"name": "test"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`},
+		},
+		"valid spec is valid": {
+			input:    `{"pipeline": {"name": "test"}, "datumTries": 1, "resourceRequests":{"disk":"256Mi", "cpu": .5}, "autoscaling":true}`,
+			expected: []string{`{"pipeline": {"name": "test"},"resourceRequests":{"disk":"256Mi", "cpu": 0.5},"datumTries": 1, "autoscaling":true}`},
+		},
+		"invalid spec is invalid": {
+			input:     `{"not_aField": 2}`,
+			expectErr: true,
+		},
+		"multiple specs work": {
+			input: `{"pipeline": {"name": "test"}}
+---
+{"pipeline": {"name": "test2"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`, `{"pipeline": {"name": "test2"}}`},
+		},
+		"lists work": {
+			input: `
+- {"pipeline": {"name": "test"}}
+- {"pipeline": {"name": "test2"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`, `{"pipeline": {"name": "test2"}}`},
+		},
+		"null works": {
+			input: `
+                                  pipeline:
+                                    name: test
+                                  resourceRequests: null`,
+			expected: []string{`{"pipeline": {"name": "test"},"resourceRequests": null}`},
+		},
+		"sequences work": {
+			input: `
+                                 pipeline:
+                                   name: test
+                                 transform:
+                                   cmd: ["foo", "bar"]`,
+			expected: []string{`{"pipeline": {"name": "test"}, "transform": {"cmd": ["foo", "bar"]}}`},
+		},
+	}
```

Comment:
Although JSON is a subset of YAML, it would be good to test the native YAML syntax for multiple pipelines (yaml `---` yaml).

Reply:
Technically that’s what the test `multiple tests work` did; I’ve updated it to specifically use YAML-only syntax for the first spec.